### PR TITLE
WIP Introduce ConstValue::ScalarPair

### PR DESCRIPTION
**Pull request by cjgillot**
Both the interpreter and codegen have dedicated paths for scalar pairs. Having one in ConstValue makes is able to represent all "immediate" values.

In order to mitigate the increase in the size of the ConstValue struct, a first commit interns it.

### CODE REVIEW

**Comment by RalfJung** [comment]
We had `ConstValue::ScalarPair` many years ago and got rid of it. I see no reason to bring it back. `ConstValue` and `OpTy` are different types for different purposes and trying to get them to have the same shape is IMO misguided.

I have laid down my vision for how `ConstValue` should be changed in https://github.com/rust-lang/rust/issues/115877.

File: compiler/rustc_middle/src/mir/interpret/value.rs [hunk 1]
```diff
@@ -28,16 +26,25 @@ pub struct ConstAlloc<'tcx> {
...
+
 /// Represents a constant value in Rust. `Scalar` and `Slice` are optimizations for
 /// array length computations, enum discriminants and the pattern matching logic.
 #[derive(Copy, Clone, Debug, Eq, PartialEq, TyEncodable, TyDecodable, Hash)]
 #[derive(HashStable, Lift)]
-pub enum ConstValue<'tcx> {
+pub enum ConstValueKind<'tcx> {
     /// Used for types with `layout::abi::Scalar` ABI.
     ///
     /// Not using the enum `Value` to encode that this must not be `Uninit`.
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/value.rs:40]: I do wonder which enum this refers to... I think it's just a very outdated comment.^^

File: compiler/rustc_middle/src/mir/interpret/value.rs [hunk 2]
```diff
@@ -50,7 +57,7 @@ pub enum ConstValue<'tcx> {
 
     /// A value not representable by the other variants; needs to be stored in-memory.
     ///
-    /// Must *not* be used for scalars or ZST, but having `&str` or other slices in this variant is fine.
+    /// Must *not* be used for scalars, scalar pairs or ZST.
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/value.rs:60]: Your `const_to_op` violates this rule for scalar pairs. Either the comment or `const_to_op` need to be updated.

File: compiler/rustc_middle/src/mir/interpret/value.rs [hunk 3]
```diff
@@ -28,16 +26,25 @@ pub struct ConstAlloc<'tcx> {
...
+pub enum ConstValueKind<'tcx> {
     /// Used for types with `layout::abi::Scalar` ABI.
     ///
     /// Not using the enum `Value` to encode that this must not be `Uninit`.
     Scalar(Scalar),
 
+    /// Used for types with `layout::abi::ScalarPair` ABI.
+    ///
+    /// Not using the enum `Value` to encode that this must not be `Uninit`.
+    ScalarPair(Scalar, Scalar),
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/interpret/value.rs:46]: We definitely don't want to have `ScalarPair` *and* `Slice`. Why would that make sense?

**Comment by Dylan-DPC** [comment]
Closing this as it is WIP and heavily bitrotted

**Pull request closed without merging**